### PR TITLE
Fix torch::deploy workaround again.

### PR DESCRIPTION
**Pull request by kumpera**
There are two issues, can't init Library unconditionally and the torch::deploy
test is wrong. Let's breakdown:

1) Can't init Library unconditionally.

We create a "DEF" Library and that never worked under torch::deploy, to the point
that it stubs out torch._meta_registrations to avoid triggering it.

We still need to create it at the module level or the whole thing breaks. Moving
the creation under the if body fixes this.

2) torch::deploy test is wrong

The support code to the original probing method was removed in August 2022.
We're now using the fact that torch._meta_registrations is stubbed out instead.

### CODE REVIEW

File: torch/__init__.py [hunk 1]
```diff
@@ -1650,3 +1650,7 @@ def _sparse_coo_tensor_unsafe(*args, **kwargs):
 
 from . import _logging
 _logging._init_logs()
+
+# Register functional collectives
+import torch.distributed._functional_ops as _functional_ops
+del _functional_ops
```
Comment (wconstab) [inline torch/__init__.py:1656]: im not sure what the point of this is, probably need a comment here

File: torch/distributed/_functional_collectives.py [hunk 2]
```diff
@@ -64,6 +62,18 @@
...
+Registration logic lives in _functional_ops.py and it's imported by torch/__init__.py.
+It's done this way to avoid issues with torch::deploy not reliably seeing our ops registered.
+
+
+** Adding a new Op:
+
+Add the op definition to _functional_ops.py::ops_defs, the format is the one torch.library expects.
+Add _{op} function to this file with matching signature. This is the op implementation.
+Add _{op}_meta function to this file with the meta kernel to be used by symbolic tracing.
+aAdd {op} high level version to this file. It's what users are expected to call.
```
Comment (wconstab) [inline torch/distributed/_functional_collectives.py:76]: typo

File: torch/distributed/_functional_ops.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+c10_lib_impl = torch.library.Library("c10d_functional", "IMPL")
+
+ops_defs = [
+    "all_reduce(Tensor self, str reduceOp, str tag, int[] ranks, int group_size) -> Tensor",
+    "wait_tensor(Tensor self) -> Tensor",
+    "all_gather_into_tensor(Tensor shard, str tag, int[] ranks, int group_size) -> Tensor",
+    "reduce_scatter_tensor(Tensor input, str reduceOp, str tag, int[] ranks, int group_size) -> Tensor",
+]
+
+def _register_ops(resolve_ops):
```
Comment (wconstab) [inline torch/distributed/_functional_ops.py:14]: why is resolve_ops passed here, it's not used?

File: torch/__init__.py (hunk 1 above)
Comment (kumpera) [reply torch/__init__.py:1656]: The point is ensuring no namespace pollution.
We want make sure that nobody can do `import torch._functional_ops as big_mistake`

Comment (wconstab) [reply torch/__init__.py:1656]: i see, yea that makes sense. you're just doing this to trigger the init for deploy, iiuc

File: torch/distributed/_functional_ops.py (hunk 3 above)
Comment (kumpera) [reply torch/distributed/_functional_ops.py:14]: Yeah, it's half done. Fixed :/

File: aten/src/ATen/core/dispatch/Dispatcher.cpp [hunk 4]
```diff
@@ -79,7 +79,7 @@ c10::optional<OperatorHandle> Dispatcher::findOp(const OperatorName& overload_na
 void Dispatcher::waitForDef(const FunctionSchema& schema) {
   using namespace std::chrono_literals;
   std::unique_lock<std::mutex> lock(mutex_);
-  bool r = cond_var_.wait_for(lock, 2s, [&]{
```
Comment (wconstab) [inline aten/src/ATen/core/dispatch/Dispatcher.cpp:82]: oh, what is this?

File: torch/distributed/_functional_ops.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    import torch.distributed._functional_collectives as fun_col
+    my_module = fun_col
```
Comment (wconstab) [inline torch/distributed/_functional_ops.py:15]: shouldn't fun_col only be imported in the `if torch.distributed.is_available() True` case?  It looks like you'll import it even if not available here.  I think you meant to delete those 2 lines above?

File: aten/src/ATen/core/dispatch/Dispatcher.cpp (hunk 4 above)
Comment (kumpera) [reply aten/src/ATen/core/dispatch/Dispatcher.cpp:82]: This is the setup that enables multiple interpreters to handle op registration.
It requires waiting for ops to pop up with a given timeout.
I'll remove this change if my latest hack works. 🤞

File: torch/distributed/_functional_ops.py (hunk 5 above)
Comment (kumpera) [reply torch/distributed/_functional_ops.py]: Import inside functions are only evaluated when the function is first called. That's why this works. :)

I removed this as I simplified the change even further.

**Review by ezyang — APPROVED** [review]
k

**Pull request closed without merging**